doc/notifications.md

```diff
@@ -1,4 +1,308 @@
 # Notifications
 
-**TODO(stevvooe)** Cover use and deployment of webhook notifications. Link to
-description in architecture documentation.
\ No newline at end of file
+> **TODO:** Link out to the architecture document on notification support.
```

Comment:
# Notifications

> **TODO:** Link out to the architecture document on notification support.

The Docker Registry Service supports sending webhook notifications in response
to registry events. A registry sends notifications in response to pushes and
pulls of both manifests and layers. Each action is serialized into an event.
Events go into a registry-internal broadcast system that queues and dispatches
events to [_Endpoints_](#endpoints).

> **TODO:** Insert diagram of event system.

## Endpoints

A registry sends notifications to _endpoints_ via HTTP requests. The registry admin must configure these endpoints. Each endpoint has isolated queues, retry configuration and http targets within a registry instance. 

When a registry detects an action, it converts the action into an event and
drops it into an inmemory queue. When the event reaches the end of the queue,
the service makes an http request to the endpoint until the request succeeds.
The service sends events serially to each endpoint but their order is not
guaranteed.

## Configuration

To configure a registry to send notifications to endpoints, you add
endpoints to its configuration, for example:

```yaml
notifications:
  endpoints:
    - name: alistener
	  url: https://mylistener.example.com/event
      headers:
        Authorization: [Bearer <your token, if needed>]
      timeout: 500ms
      threshold: 5
      backoff: 1s
```

This example configures the registry with an endpoint to send events to
"https://mylistener.example.com/event", with the header "Authorization: Bearer
<your token, if needed>". The request timeouts after 500 milliseconds. If
5 failures happen consecutively, the registry pauses for 1 second before
trying again.

For details on the notification configuration, please see the [configuration documentation](configuration.md#notifications).

A properly configured endpoint results in a log message from the registry
upon startup:

```
INFO[0000] configuring endpoint alistener (https://mylistener.example.com/event), timeout=500ms, headers=map[Authorization:[Bearer <your token if needed>]]  app.id=812bfeb2-62d6-43cf-b0c6-152f541618a3 environment=development service=registry
```

> TODO: Steve what does failure look like? 

## Events

Events have a well-defined JSON structure and are sent as the body of
notification requests. A registry sends one or more events in a structure called
an envelope. Each event has a unique id that clients can use to eliminate
duplicates from incoming requests, if required. Along with that, the response
includes an `action` and `target` that describe how and what the event changed.

The following is an example of a JSON event, sent in response to a manifest push:

```json
{
   "id": "asdf-asdf-asdf-asdf-0",
   "timestamp": "2006-01-02T15:04:05Z",
   "action": "push",
   "target": {
      "mediaType": "application/vnd.docker.distribution.manifest.v1+json",
      "length": 1,
      "digest": "sha256:0123456789abcdef0",
      "repository": "library/test",
      "url": "http://example.com/v2/library/test/manifests/latest"
   },
   "request": {
      "id": "asdfasdf",
      "addr": "client.local",
      "host": "registrycluster.local",
      "method": "PUT",
      "useragent": "test/0.1"
   },
   "actor": {
      "name": "test-actor"
   },
   "source": {
      "addr": "hostname.local:port"
   }
}
```

The fields available in an event are described in detail in the
[godoc](http://godoc.org/github.com/docker/distribution/notifications#Event).

> **TODO:** Let's break out the fields here rather than rely on the godoc.

## Envelope

The envelope contains one or more events with the following json structure:

```json
{
	"events": [ ... ],
}
```

While events may be sent in the same envelope, the set of events within that
envelope may be unrelated. For example, the registry may choose to
group events and send them in the same envelope to reduce the total
number of requests.

An envelope's mediatype is `application/vnd.docker.distribution.events.v1+json` and it is set on the request coming to an endpoint, for example:

```json
GET /callback
Host: application/vnd.docker.distribution.events.v1+json
Authorization: Bearer <your token, if needed>
Content-Type: application/vnd.docker.distribution.events.v1+json

{
   "events": [
      {
         "id": "asdf-asdf-asdf-asdf-0",
         "timestamp": "2006-01-02T15:04:05Z",
         "action": "push",
         "target": {
            "mediaType": "application/vnd.docker.distribution.manifest.v1+json",
            "length": 1,
            "digest": "sha256:0123456789abcdef0",
            "repository": "library/test",
            "url": "http://example.com/v2/library/test/manifests/latest"
         },
         "request": {
            "id": "asdfasdf",
            "addr": "client.local",
            "host": "registrycluster.local",
            "method": "PUT",
            "useragent": "test/0.1"
         },
         "actor": {
            "name": "test-actor"
         },
         "source": {
            "addr": "hostname.local:port"
         }
      },
      {
         "id": "asdf-asdf-asdf-asdf-1",
         "timestamp": "2006-01-02T15:04:05Z",
         "action": "push",
         "target": {
            "mediaType": "application/vnd.docker.container.image.rootfs.diff+x-gtar",
            "length": 2,
            "digest": "tarsum.v2+sha256:0123456789abcdef1",
            "repository": "library/test",
            "url": "http://example.com/v2/library/test/manifests/latest"
         },
         "request": {
            "id": "asdfasdf",
            "addr": "client.local",
            "host": "registrycluster.local",
            "method": "PUT",
            "useragent": "test/0.1"
         },
         "actor": {
            "name": "test-actor"
         },
         "source": {
            "addr": "hostname.local:port"
         }
      },
      {
         "id": "asdf-asdf-asdf-asdf-2",
         "timestamp": "2006-01-02T15:04:05Z",
         "action": "push",
         "target": {
            "mediaType": "application/vnd.docker.container.image.rootfs.diff+x-gtar",
            "length": 3,
            "digest": "tarsum.v2+sha256:0123456789abcdef2",
            "repository": "library/test",
            "url": "http://example.com/v2/library/test/manifests/latest"
         },
         "request": {
            "id": "asdfasdf",
            "addr": "client.local",
            "host": "registrycluster.local",
            "method": "PUT",
            "useragent": "test/0.1"
         },
         "actor": {
            "name": "test-actor"
         },
         "source": {
            "addr": "hostname.local:port"
         }
      }
   ]
}
```

## Responses

The registry is fairly accepting of the response codes from endpoints. If an
endpoint responds with any `2XX` or `3XX` response code (after following
redirects), the registry considers the message delivered and discarded.

As well, we recommended that endpoints should be similarly accepting of incoming
responses. While the format of event envelopes are standardized by media type,
any "pickiness" about validation can backup the event queue on the registry.

## Monitoring

The endpoint states are reported via the `debug/vars` http interface which is typically configured to `http://localhost:5001/debug/vars`. Information such as
configuration and metrics are available by endpoint.

The following provides and example of a few endpoints that have experience
several failures and have since recovered:

```json
"notifications":{
   "endpoints":[
      {
         "name":"local-8082",
         "url":"http://localhost:5003/callback",
         "Headers":{
            "Authorization":[
               "Bearer \u003can example token\u003e"
            ]
         },
         "Timeout":1000000000,
         "Threshold":10,
         "Backoff":1000000000,
         "Metrics":{
            "Pending":76,
            "Events":76,
            "Successes":0,
            "Failures":0,
            "Errors":46,
            "Statuses":{

            }
         }
      },
      {
         "name":"local-8083",
         "url":"http://localhost:8083/callback",
         "Headers":null,
         "Timeout":1000000000,
         "Threshold":10,
         "Backoff":1000000000,
         "Metrics":{
            "Pending":0,
            "Events":76,
            "Successes":76,
            "Failures":0,
            "Errors":28,
            "Statuses":{
               "202 Accepted":76
            }
         }
      }
   ]
}
```

If using notification as part of a larger application, it is _critical_ to
monitor the `Pending` value in each response. Its value reflects the size of the
endpoint queues. If failures or queue sizes increase, the change can indicate a
larger problem.

The logs are also a valuable resource for monitoring problems. A failing
endpoint can lead to messages similar to the following:

```
ERRO[0340] retryingsink: error writing events: httpSink{http://localhost:5003/callback}: error posting: Post http://localhost:5003/callback: dial tcp 127.0.0.1:5003: connection refused, retrying
WARN[0340] httpSink{http://localhost:5003/callback} encountered too many errors, backing off
```

The above indicates that several errors caused the queue to backoff and the registry is waiting before retrying.

## Considerations

Currently, the queues are in memory. So, endpoints should be _reasonably
reliable_. Queues attempt to make a best-effort to send the messages but if
an instance is lost, it may drop messages. If an endpoint goes down, take care to ensure that the registry instance is not terminated before
the endpoint comes back up or messages will be lost.

You can mitigate message loss by running endpoints in close proximity to the registry instances. For example, you could run an endpoint that pages to disk and then forwards a request to provide better durability.

The notification system is designed around a series of interchangeable _sinks_
that you can wire up to achieve interesting behavior. If this system doesn't
provide acceptable guarantees, adding a transactional `Sink` to the registry
is a possibility. Although, such a sink can effect request service time.
Please see the
[godoc](http://godoc.org/github.com/docker/distribution/notifications#Sink) for
more information.